Adaptive Trust Region Algorithm

peoe opened the pull request:
This PR implements the BFGS and TR algorithms as previously done by @TiKeil in his [repo](https://github.com/TiKeil/NCD-corrected-TR-RB-approach-for-pde-opt). Additionally, the parameter space tests are simplified to use a hypothesis strategy, because the BFGS subtest required data which can just be reused in parameter testing.

With regard to the citations: Yue & Meerbergen published the basic idea for the TR termination criterion, but features like enlarging of the trust radius have mainly been mentioned in Tim's paper, so I added it as well. The exact choice may be up for debate.

As soon as #1796 is merged, I'll be able to implement a demo highlighting the combined use of a bilinear output functional and the TR algorithmed used here.

TiKeil commented on src/pymortests/algorithms/tr.py:
```diff
@@ -0,0 +1,30 @@
+
+    # successful run
+    _, data = trust_region(parameter_space, reductor, radius=.1, initial_guess=initial_guess)
+    assert len(data['mus']) == 22
```
This would be a very disappointing number of outer iterations. Instead, setting the default for `atol` in the TR to `1e-16` gives only 4 outer iterations. It is very typical for such TR algorithms that they are very sensitive to tolerance choices.

peoe replied:
Right, playing around with the values a bit I noticed that choosing `atol <= 1e-6` did not significantly change the convergence behaviour of the TR for this example. Taking a page from `newton.py`s book, I'll use the `atol = 0.` from there.

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,292 @@
+def trust_region(parameter_space, reductor, initial_guess=None, beta=.95, radius=1.,
+                 shrink_factor=.5, miniter=0, maxiter=100, miniter_subproblem=0, maxiter_subproblem=100, tol=1e-6,
+                 radius_tol=.75, atol=1e-6, tol_sub=1e-6, line_search_params=None, stagnation_window=3,
+                 stagnation_threshold=np.inf, return_stages=False, return_subproblem_stages=False):
```
I think the default choices of e.g. the tolerances in general should be discussed and evaluated further. Moreover, there should be a different `tol_sub` for the bfgs and for the TR. See the comment in the test.

peoe replied:
I'm a bit confused by this comment: `tol_sub` is only used in the BFGS part of the algorithm. Do you mean that `atol` and `tol_sub` should differ from one another?

TiKeil replied:
I am sorry, my bad. I was referring to `tol` and `tol_sub`, which already differ. So forget about my comment.

TiKeil commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,309 @@
+    if error_aware:
+        assert beta is not None and radius is not None
+
+        def error_aware_line_search_criterion(f, starting_point, alpha, direction, grad, slope):
```
I understand that this is the most flexible way of allowing for an additional line search criterion. In my opinion, however, it should be enough to only have `new_mu` and `current_value`, (which is available in the line search) as an argument.

TiKeil commented on src/pymor/algorithms/line_search.py:
```diff
@@ -56,15 +60,22 @@ def armijo(f, starting_point, direction, grad=None, initial_value=None,
         current_value = f(starting_point + alpha * direction)
         # Check the Armijo condition
-        if current_value < initial_value + alpha * beta * slope:
+        if (current_value < initial_value + alpha * beta * slope
+                or additional_criterion(f, starting_point, alpha, direction, grad, slope)):
```
```suggestion
        if (current_value < initial_value + alpha * beta * slope
                or additional_criterion(current_value, new_mu)):
```
and `new_mu` defined as in line 73

TiKeil reviewed (changes requested):
Thanks for the PR @peoe !

Regarding the overall design, I am quite happy (I have not looked into all the technical details yet)

I think the logging output should definitely be improved. It is a little bit difficult to follow the algorithm through its stages. For instance, I think it is very important to document the `mu_clip_norm`. Speaking of which, I think this is a strange name for an FOC-based criterion. Also, I do not think that `mu_norm` is particularly interesting as output.
Moreover, it is of interest how many iterations the BFGS took and how many iterations the line search took.

It would also be nice to have something like a small final report like `TR converged due to [...] after [..] iterations with a reduced basis size of [..] and a TR radius of [..].` something similar (similar to the greedy).

We should also further think about the default tolerances. For this I would propose to have a look at my repo again and start with these (see comment below).

peoe commented:
I have now cleaned up the output quite a bit, which should make it much easier to follow. The number of iterations of each subroutine is now also stored in the returned data.

With regards to the default tolerances I ran a little script and figured out that the number of outer TR iterations might even be reduced to 3 depending on the choice of `tol_sub` and `atol`. Do you have any opinions on how to choose these values?
[test_tr_params.txt](https://github.com/pymor/pymor/files/10472367/test_tr_params.txt) &emsp; [test_tr_params_output.txt](https://github.com/pymor/pymor/files/10472368/test_tr_params_output.txt)
(The script ran in ~35 second for me; the output also contains a comparison of FOM and ROM output which I removed in the `test_tr_params_output` file.)

TiKeil commented:
> I have now cleaned up the output quite a bit, which should make it much easier to follow. The number of iterations of each subroutine is now also stored in the returned data.
> 
> With regards to the default tolerances I ran a little script and figured out that the number of outer TR iterations might even be reduced to 3 depending on the choice of `tol_sub` and `atol`. Do you have any opinions on how to choose these values? [test_tr_params.txt](https://github.com/pymor/pymor/files/10472367/test_tr_params.txt)   [test_tr_params_output.txt](https://github.com/pymor/pymor/files/10472368/test_tr_params_output.txt) (The script ran in ~35 second for me; the output also contains a comparison of FOM and ROM output which I removed in the `test_tr_params_output` file.)

Thanks for the script and the results. It shows that the choice of tolerances matters. A few comments from an optimization perspective and their impact on the outer iterations:

- `tol`: It is clear that the number of outer iterations heavily depends on this FOC tolerance because it simply means how far we aim to converge to the local minimum. `tol` should never be `0` unless we are not interested in the FOC condition. I think `1e-6` would be a good default.
- `tol_sub` determines the FOC termination for the inner problem. Since the inner problem is only on ROM level and thus very cheap, `tol_sub` is typically chosen much smaller than `tol`. It should also not be set to `0` unless we do not want an FOC condition in the inner iteration. Setting `tol_sub` too high hurts the number of outer iterations because the inner iterations are terminated too early, which does not make sense, in my opinion. I would go for something like `1e-8` or even lower as default.
- `a_tol` should generally be considered a "safety tolerance" since (if a local minimum exists) the FOC conditions should do the job. If something weird happened in the optimization or if `tol` and/or `tol_sub` have been set too small (or `0`), `a_tol` prevents the algorithm from stagnating. Setting`a_tol` too high will have the same effect as setting `tol_sub` too high: the inner problem can not converge, although it is very cheap. For `a_tol`, I would default something between `1e-12` or `1e-16`. However, `0` should never be used.

Do you agree?

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,306 @@
+
+        # generate a first rom if none was given
+        self.fom = reductor.fom
+        self.rom = None
+        if reductor._last_rom is None:
+            initial_basis = reductor.fom.solve(initial_guess)
+            reductor.extend_basis(initial_basis)
+            reductor.reduce()
+        self.rom = reductor._last_rom
```
```suggestion
        self.rom = reductor.reduce()
```
and put the first enrichment in the TR algorithm for clarity?

peoe replied:
I wasn't quite sure about this. My thought was that adding the initial "extension" of the surrogate RB into the TR would add (unnecessary) clutter: after all, if the end user just wants to understand the TR algorithm they would not need to know that the surrogate is always enriched for the initial guess upon initialization without a reduced basis.

TiKeil replied:
Would it be a better idea to use the `extend` method here? Then the initial extension fits to the other extensions and the user only needs to adapt the method.

peoe replied:
I have now reduced the code in the `TRSurrogate` to

```python
# generate a first rom if none was given
self.fom = reductor.fom
self.extend(initial_guess)
self.accept()
```

this still keeps the initial extension outside of the TR but has to take the detour through `accept()` because of the TR logic linked to extensions.

Please let me know if this wasn't what you had intended.

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,306 @@
+            if iteration >= miniter:
+                if first_order_criticity < tol:
+                    logger.info(
+                        f'TR converged in {iteration} iterations because first order criticity tolerance of {tol}' \
```
```suggestion
                        f"TR converged in {iteration} iterations because first order criticity tolerance of {tol}"
```
or give the surrogate a method that returns the basis length.

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,306 @@
+                # ignore division-by-zero warnings when solution_norm or output is zero
+                warnings.filterwarnings('ignore', category=RuntimeWarning)
+                logger.info(f'it:{iteration} '
+                            f'norm:{mu_norm:.3e} '
```
```suggestion
                            f'foc:{first_order_criticity:.3e} '
```

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,306 @@
+            old_mu = mu.copy()
+            compare_output = surrogate.rom_output(mu)
+
+            with logger.block('Solving subproblem for mu {mu} with BFGS...'):
```
```suggestion
            with logger.block(f'Solving subproblem for mu {mu} with BFGS...'):
```

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,306 @@
+
+@defaults('beta', 'radius', 'shrink_factor', 'miniter', 'maxiter', 'miniter_subproblem', 'maxiter_subproblem',
+          'tol', 'radius_tol', 'atol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def trust_region(parameter_space, reductor, initial_guess=None, beta=.95, radius=1.,
```
```suggestion
def trust_region(parameter_space, reductor, initial_guess=None, beta=.95, radius=.1,
```
`radius=1.` means that we trust the ROM 100%. Having 10% is generally more feasible (although 100% gives less outer iterations in the test we currently have).

TiKeil reviewed (changes requested):
Found a few minor things.

peoe commented:
I've moved the additional line search criterion to the TR level because the convergence check there used the same logic. The function is now passed to BFGS using an additional argument.

TiKeil commented:
We should think about what happens with the `parameter_space` argument if we do not have constraints on the parameter space. Maybe we should have it optional and default with a full parameter space.

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,317 @@
+
+            # solve the subproblem using bfgs
+            old_mu = mu.copy()
+            compare_output = surrogate.rom_output(mu)
```
this should be the AGC point, e.g. the first iteration of the BFGS

TiKeil commented:
Apart from minor things, I think this is ready to be reviewed regarding the overall design.

HenKlei commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,307 @@
+    return active, inactive
+
+
+def project_hessian(hessian, direction, active, inactive):
+    """Project the Hessian matrix using an active and an inactive set.
```
The function name and also the description do not seem to fit to what the function does. It does not return a "projected Hessian", but the projection of the Hessian applied to the direction.

peoe replied:
I'll take a hint from Tim's repo and name it `compute_hessian_action_matrix`.

HenKlei reviewed (changes requested):
I will continue looking through this tomorrow or next week...

HenKlei commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,307 @@
+        foc_norms.append(first_order_criticity)
+
+        # set new active inactive threshhold
+        eps_update_mu = mu - gradient
+        eps = np.linalg.norm(eps_update_mu - mu)
```
```suggestion
        eps = np.linalg.norm(gradient)
```

HenKlei commented on src/pymortests/algorithms/tr.py:
```diff
@@ -0,0 +1,30 @@
+
+    # successful run
+    _, data = trust_region(parameter_space, reductor, radius=.1, initial_guess=initial_guess)
+    assert len(data['mus']) == 4
```
See my comment on the check in `bfgs.py`.

HenKlei commented on src/pymortests/algorithms/bfgs.py:
```diff
@@ -0,0 +1,42 @@
+
+    # successful run
+    _, data = bfgs(fom, parameter_space, initial_guess=initial_guess)
+    assert len(data['mus']) == 13
```
I think this test should be extend by also checking that the algorithm found the optimum - either by checking that the resulting mu is close to the optimal parameter value or by checking that the objective function value is close to the optimum.

HenKlei reviewed (changes requested):
`tr.py` is still missing, I am going to look into that next week.

HenKlei commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,307 @@
+    data['iterations'] = iteration
+    data['line_search_iterations'] = np.array(line_search_iterations)
+    if return_stages:
+        data['stages'] = np.array(stages)
```
I think this will not work, since `stages` contains parameters. I think you can remove either `mus` or `stages` since they are the same, if I am not mistaken.

HenKlei reviewed (dismissed):
Apart from minor remarks, this looks very good to me!

I think the structure is very nice and the algorithm is written very clearly. I also do not see any mathematical issues here, it all looks very solid.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,307 @@
+    maxiter
+        Fail if the iteration count reaches this value without converging.
+    atol
+        Finish when the absolute error measure is below this threshold.
```
Right, I got confused there by naming the error terms `..._diff`.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,307 @@
+        else:
+            step_size, line_search_iteration = armijo(output, mu, direction, grad=gradient,
+                               initial_value=current_output, **(line_search_params or {}))
+        line_search_iterations.append(line_search_iteration + 1)
```
I thought that the line search somehow computed the number of iterations that each run took with an offset of 1 compared to the BFGS/TR, but that assumption was wrong.

TiKeil commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,311 @@
+
+
+@defaults('miniter', 'maxiter', 'rtol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def bfgs(model, parameter_space, initial_guess=None, miniter=0, maxiter=100, rtol=1e-16,
```
Without the `error_aware`-part, I completely agree that it would not be necessary to have it.
The main thing why this BFGS is different is the (very unusual) additional line search criterion which is called `error_criterion` below in the code. Back in the days, I did not manage to incorporate this into the `scipy` code although it is basically only a function call.

However, it might be that the criterion nowadays can be incorporated to the BFGS within the `constraints` argument in `scipy.optimize.minimize`. @peoe, have you looked into this? Might be worth a check. Or do you see another reason why we write our own BFGS?

peoe replied:
As far as I can tell, SciPy only utilizes the constraints parameter for ["Only for COBYLA, SLSQP and trust-constr."](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize). So unless we were to perform some intrusive things during runtime or always manually treat the data after the algorithm to fit into our framework I don't see how we might incorporate the additional criterion with scipy...

Besides avoiding potential dependency troubles further down the line, another minor argument is that if one were to extend the TR/BFGS code (e. g. to the non-conforming dual setting that Tim covered or for parabolic problems, ...), then providing a standalone implementation will be much simpler to work with.

TiKeil replied:
I just had another look at SciPy. It definitely does not work with what is implemented there. The inequality constraint that we would need can only be treated within the 'SLSQP' which heavily iterates around the boundary of the trust-region (which we do not allow in our algorithm).

Of course contributing to SciPy is an option, but I honestly do not know any application where it makes sense to terminate the sub-problem earlier instead of iterating across the boundary. Especially because we are not interested in implementing the TR algorithm but only the error-aware sub-problem. The feature request would go down to changing the [armijo line search](https://github.com/scipy/scipy/blob/9aee554b9f87d53eb32fd2b0318c45439c9d9950/scipy/optimize/_linesearch.py#L669) in SciPy which I do not think they would be much interested in.

For reference: I used the following lines of code for the subproblem in `pymor/algorithms/tr.py`

```python 
with logger.block(f'Solving subproblem for mu {mu} with BFGS...'):
    # BFGS with error_aware criterion as we use it now ! 
    mu, sub_data = bfgs(
        surrogate.rom, parameter_space, initial_guess=mu, miniter=miniter_subproblem,
        maxiter=maxiter_subproblem, rtol=rtol, tol_sub=tol_sub,
        line_search_params=line_search_params, stagnation_window=stagnation_window,
        stagnation_threshold=stagnation_threshold, error_aware=True,
        error_criterion=error_aware_line_search_criterion, beta=beta, radius=radius)

    # BFGS without error_aware criterion
    mu_, sub_data_ = bfgs(
        surrogate.rom, parameter_space, initial_guess=old_mu, miniter=miniter_subproblem,
        maxiter=maxiter_subproblem, rtol=rtol, tol_sub=tol_sub,
        line_search_params=line_search_params, stagnation_window=stagnation_window,
        stagnation_threshold=stagnation_threshold, error_aware=False,
        error_criterion=error_aware_line_search_criterion, beta=beta, radius=radius)
    
    # BFGS in scipy without error_aware criterion
    def objective_functional(mu_in):
        return surrogate.rom.output(mu_in)[0, 0]

    def gradient_of_functional(mu_in):
        return surrogate.rom.output_d_mu(surrogate.rom.parameters.parse(mu_in), return_array=True)
    
    result = minimize(objective_functional,
                      old_mu,
                      method='L-BFGS-B',
                      jac=gradient_of_functional,
                      bounds=(parameter_space.ranges['diffusion'], parameter_space.ranges['diffusion']),
                      options={'ftol': rtol, 'gtol': tol_sub})

    def error_aware_line_search_criterion_for_scipy(new_mu):
        current_value = surrogate.rom.output(new_mu)[0, 0]
        output_error = surrogate.estimate_output_error(new_mu)
        return output_error / abs(current_value)
    
    # SLSQP from scipy with error_aware criterion
    from scipy.optimize import NonlinearConstraint
    const = NonlinearConstraint(error_aware_line_search_criterion_for_scipy, 0, beta * radius)

    result_ = minimize(objective_functional,
                      old_mu,
                      method='SLSQP',
                      jac=gradient_of_functional,
                      bounds=(parameter_space.ranges['diffusion'], parameter_space.ranges['diffusion']),
                      options={'ftol': rtol, 'gtol': tol_sub},
                      constraints=const)
```
Looking at the results in this specific example:
Our self-written BFGS takes 2 iterations before the boundary is reached. The SLSQP instead takes 36 iterations until it finishes.
This already shows that iterating across the boundary is not feasible for our algorithm.

Maybe as a side note: In order to make clear that the BFGS is special in that regard, I would call it `error_aware_BFGS` or something similar.

pmli replied:
Ok, then I think it would be good to add an explanation in the `bfgs` docstring about what makes it special. Renaming the function also sounds good.

sdrave replied:
I must say that I do not completely understand what is available in scipy compared to what is implemented here. Regarding BFGS, there also is a `BFGS` `HessianUpdateStrategy` which one might consider to use. Then, there are also implementations of trust region methods in scipy. In particular, there is also the (somewhat private) `_minimize_trust_region` method, which you can pass a `subproblem`. Have you looked into that?

TiKeil replied:
Thanks @sdrave  for pointing into this direction.

AFAIK, `BFGS-B` (`B` for bounds) is only implemented within the `L-BFGS-B` (`L` for limited-memory) variant. The `L-BFGS-B` is Fortran code which I do not intend to look into. Also: I do not see the point of using it, since these are only 3 relatively simple methods in our case.

Regarding the `_minimize_trust_region` methods in `scipy`, I want to remind you that our method is mathematically very different to what is done in `_minimize_trust_region` and can not be used. Moreover, a generalization to our liking would probably be of low interest to `scipy`. Apart from this, we would have to implement our own `subproblem` (which then would have code that is currently in our `BFGS` method).

Just to make this point clear: The Trust-Region methods in scipy all have in common that the `trust-region` is determined by a metric distance, not an actual `error_estimator`. Moreover, the cheap model function for the subproblem is constructed from scratch, whereas, in our method, the model function is reused but updated.

sdrave replied:
Ok. Although I think that it's still kind of a shame that we have to reimplement everything, I see that it is probably not feasible to get the scipy code adapted to our needs.

> Just to make this point clear: The Trust-Region methods in scipy all have in common that the trust-region is determined by a metric distance, not an error_estimator. Moreover, the cheap model function for the subproblem is constructed from scratch for each outer iteration, whereas, in our method, the model function is reused but updated.

Some summary like that should definitely go into the docstring.

TiKeil replied:
> Ok. Although I think that it's still kind of a shame that we have to reimplement everything, I see that it is probably not feasible to get the scipy code adapted to our needs.

I do not understand the "shame". It is a completely different method that is no use for `scipy` unless they want to do model reduction, which is our thing right?

> 
> > Just to make this point clear: The Trust-Region methods in scipy all have in common that the trust-region is determined by a metric distance, not an error_estimator. Moreover, the cheap model function for the subproblem is constructed from scratch for each outer iteration, whereas, in our method, the model function is reused but updated.
> 
> Some summary like that should definitely go into the docstring.

I agree !

sdrave replied:
> I do not understand the "shame". It is a completely different method that is no use for `scipy` unless they want to do model reduction, which is our thing right?

I wouldn't agree that the method is 'completely' different, and I would assume that you can think of other surrogates, which do not come from MOR as we understand ist. However, let's have that discussion somewhere else.

I'll try to do a full review today or tomorrow.

peoe replied:
To summarize the changes I've just added (I hope I didn't skip over any feedback):

peoe replied:
To summarize the changes I've just added (I hope I didn't miss any of your feedback):
* renamed `bfgs` to `error_aware_bfgs`
* added a remark to the BFGS docstring about terminating close to the TR boundary, contrasting the `scipy.optimize` behaviour
* added a remark to the TR docstring about computation of TR conditions and construction of the model surrogate

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    active = np.array([])
+
+    for key, dim in parameter_space.parameters.items():
+        indices = np.zeros(dim)
+        low_range = parameter_space.ranges[key][0] * np.ones(dim)
+        high_range = parameter_space.ranges[key][1] * np.ones(dim)
+        eps = epsilon * np.ones(dim)
+        indices[np.where(np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))] = 1
```
Using broadcasting, can't you replace this by something like

```python
low_range, high_range = parameter_space.ranges[key]
indices = np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        The set of inactive parameter indices.
+    """
+    mu = mu if isinstance(mu, Mu) else parameter_space.parameters.parse(mu)
+    active = np.array([])
```
It would probably be more efficient to collect all individual arrays in a Python list and then create the concatenated array from this list of arrays.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        indices[np.where(np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))] = 1
+        active = np.concatenate((active, indices))
+
+    inactive = np.ones(active.shape) - active
```
Using boolean indices as suggested above, one can simply write:
```suggestion
    inactive = ~active
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    Returns
+    -------
+    active
+        The set of active parameter indices.
```
You are note returning the set of indices but the corresponding mask.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    return active, inactive
+
+
+def compute_hessian_action_matrix(hessian, direction, active, inactive):
```
```suggestion
def _compute_hessian_action_matrix(hessian, direction, active, inactive):
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    hessian
+        The current Hessian matrix.
+    mu
+        The current `mu` parameter.
```
See comment above.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        The next BFGS Hessian matrix iterate.
+    """
+    gradient_difference = gradient - old_gradient
+    mu_difference = mu - old_mu
```
So `mu` and `old_mu` apparently need to be a flattened `Mu` here. Should be mentioned in the docstring.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+
+    for an output functional depending on a box-constrained `mu` using the BFGS method.
+
+    In contrast to `scipy.optimize` with the `L-BFGS-B` methods, this BFGS implementation is
```
```suggestion
    In contrast to :func:`scipy.optimize.minimize` with the `L-BFGS-B` methods, this BFGS implementation is
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+
+    In contrast to `scipy.optimize` with the `L-BFGS-B` methods, this BFGS implementation is
+    explicitly designed to work with an error estimator. In particular, this implementation
+    terminates if the higher level TR boundary from `pymor.algorithms.tr` is reached instead of
```
```suggestion
    terminates if the higher level TR boundary from :mod:`pymor.algorithms.tr` is reached instead of
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    Parameters
+    ----------
+    model
+        The |Model| used for the optimization.
```
No model is mentioned in the problem definition above. Also, is there really a need to have a `J` here, which comes from a model?

peoe replied:
The `model` basically serves as a convenient way to:
- provide the output functional $J$,
- compute output derivatives, and
- parse parameter values.

In particular the last two are of interest, with the parameter parsing serving as an assured proxy in case the user did not specify a `ParameterSpace`.

We were thinking about making this function independent of a `model`, however there were some problems with that approach, hence we stuck with the `model`.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    rtol
+        Finish when the relative error measure is below this threshold.
+    tol_sub
+        Finish when the clipped parameter error measure is below this threshold.
```
Please define 'clipped parameter error measure'.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        See `stagnation_window`.
+    error_aware
+        If `True`, perform an additional error aware check during the line search phase.
+        Intended for use with the trust region algorithm. Requires the parameters `beta`
```
Is there documentation somewhere, what to pass for the additional parameters in this case?

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    data
+        Dict containing the following fields:
+
+            :mus:                       `list` of parameters after each iteration.
```
see above.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    ------
+    BFGSError
+        Raised if the BFGS algorithm failed to converge.
+    """
```
Had to stop review here.

sdrave reviewed (changes requested):
This is a detailed review up to the location marked by my comment. Hopefully will be able to contiue tomorrow.

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        active, inactive = get_active_and_inactive_sets(parameter_space, mu, eps)
+
+        # compute update to mu
+        if sum(inactive) == 0.:
```
```suggestion
        if np.all(active):
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        else:
+            direction = compute_hessian_action_matrix(hessian, -gradient, active, inactive)
+
+        if error_aware:
+            step_size, line_search_iteration = armijo(
+                output, mu, direction, grad=gradient, initial_value=current_output,
+                additional_criterion=error_criterion, **(line_search_params or {}))
+        else:
+            step_size, line_search_iteration = armijo(output, mu, direction, grad=gradient,
+                               initial_value=current_output, **(line_search_params or {}))
```
```suggestion
        step_size, line_search_iteration = armijo(
            output, mu, direction, grad=gradient, initial_value=current_output,
            additional_criterion=error_criterion if error_aware else None, 
            **(line_search_params or {}))
```

sdrave commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        old_output = current_output.copy()
+        mu += step_size * direction
+        mu = parameter_space.clip(mu).to_numpy()
+        current_output = output(mu)
```
```suggestion
        current_output = output(mu).item()
```

peoe replied:
I don't think we actually want this. If I deitemize the value, `.copy()` no longer works because that's a `NumPy` method. Also, this code works fine (for me) without `.item()`, and later generalization to higher dimensional outputs would again break at this point.

sdrave commented on src/pymor/algorithms/line_search.py:
```diff
@@ -35,6 +37,8 @@ def armijo(f, starting_point, direction, grad=None, initial_value=None,
+        A `callable` used as an additional termination criterion during the step length computation.
 
     Returns
     -------
```
Document additional return value.

sdrave commented on src/pymor/algorithms/line_search.py:
```diff
@@ -56,15 +60,25 @@ def armijo(f, starting_point, direction, grad=None, initial_value=None,
-    if grad:
-        slope = min(grad.inner(direction), 0.0)
+    if grad is not None:
+        if isinstance(grad, np.ndarray):
```
I think it's ok to allow both VectorArrays an numpy arrays here, but the docstring needs to be updated.

sdrave reviewed (changes requested):
I have finished now reviewing the bfgs algorithm. Since I already made so many comments, I will defer reviewing the tr algorithm to when we have finished discussing bfgs.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+            if output_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for output difference reached. Converged.')
+                break
+            if mu_diff < rtol:
```
I don't think that two different parameters will matter here if the tolerances are both chosen small enough. On the other hand this would be pretty easy to add, so we might as well, right?

TiKeil replied:
Two tolerances certainly don't hurt and increases user-input flexibility.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+            if (iteration >= stagnation_window + 1 and not stagnation_threshold == np.inf
+                    and all(np.isclose(
+                        [max(update_norms[-stagnation_window - 1:]), min(update_norms[-stagnation_window - 1:])],
+                        update_norm, atol=stagnation_threshold))):
```
I changed it to a relative threshold now. The check here works differently than in `newton` though, because we don't want to check that the error is reduced, but rather that the update of the parameters stalls out.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+    output = lambda m: model.output(m)[0, 0]
+
+    if parameter_space is None:
+        parameter_space = model.parameters.space(-np.inf, np.inf)
```
I just used `np.inf` so as not to assume to much about the unspecified parameter space. I'll change it to $(-1, 1)$ and issue a warning from the logger instead.

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        parameter_space = model.parameters.space(-np.inf, np.inf)
+
+    if initial_guess is None:
+        initial_guess = parameter_space.sample_randomly(1)[0]
```
This should be fixed by the (-1, 1) fix from above.

TiKeil commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,320 @@
+        assert error_criterion is not None and beta is not None and radius is not None
+        assert callable(error_criterion)
+
+    gradient = model.parameters.parse(model.output_d_mu(mu)).to_numpy()
```
I agree, this is not ideal. I would acutally vote for a `numpy`-array, but a `Mu` instance sounds also reasonable.

pmli commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,319 @@
+        |parameter values| `mu`. Otherwise a |ParameterSpace| with lower bound -1
+        and upper bound 1.
+    initial_guess
+        If not `None`, a |Mu| instance containing an initial guess for the solution `mu`.
```
```suggestion
        If not `None`, |parameter values| containing an initial guess for the solution `mu`.
```

pmli commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,332 @@
+        |parameter values| `mu`. Otherwise a |ParameterSpace| with lower bound -1
+        and upper bound 1.
+    initial_guess
+        If not `None`, a |Mu| instance containing an initial guess for the solution `mu`.
```
```suggestion
        If not `None`, |parameter values| instance containing an initial guess for the solution `mu`.
```

pmli reviewed (commented):
> There seems to be an [actual error](https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/1615416#L2914) in the docs.

Indeed.

sdrave commented:
@peoe, @TiKeil, is this now ready for review again?

peoe commented:
@sdrave This PR is now ready for review; sorry for the delay.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+    return mu, data
+
+
+class TRSurrogate(BasicObject):
```
Somehow the name `Surrogate` does not seem quite fitting. I would not expect a surrogate expect to have methods like `fom_output` or `fom_gradient`. Is there another option for a name?
Actually, I am wondering whether it would make more sense to actually remove these methods and have the fom as a parameter of the tr algorithms. Would make sense to me that I need to give the algorithm a fom and a surrogate.

TiKeil replied:
I agree with the latter. Probably the `surrogate` should not have fom methods.

sdrave replied:
Maybe I misunderstand what you are trying to achieve here. But, what I have in mind is to remove the fom from the surrogate altogether. (Of course, depending on the impl of the surrogate, there might be a fom in the background, but that should not be part of the interface.) Instead, the fom would be passed as an additional argument to `trust_region`. The fom evaluations could be counted where they occur.

peoe replied:
Oh right, I now pass the FOM directly to the `trust_region` method as well as a method argument if we need it within the surrogate (except for the `__init__` where I use the reductor as a work around instead.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+            self.new_reductor = deepcopy(self.reductor)
+            try:
+                self.new_reductor.extend_basis(U_h_mu)
+            except Exception:
```
Please catch a more specific exception.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+
+
+class SimpleTRSurrogate(TRSurrogate):
+    """Surrogate for the :func:`trust_region` using a simple linear error estimator.
```
What is a 'linear error estimator'?

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+    reductor
+        The reductor used to generate the reduced order models and estimate the output error.
+    initial_guess
+        The `parameter values` instance containing an initial guess for the optimal parameter value.
```
See above.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+                pass
+            self.new_rom = self.new_reductor.reduce()
+
+    def estimate_output_error(self, mu):
```
Same here ..

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+        The `parameter values` instance containing an initial guess for the optimal parameter value.
+    """
+
+    def extend(self, mu):
```
Feels a bit strange that this code is in here and not the reductor or so ..

sdrave reviewed (changes requested):
So I finally managed to do another round of review, this time also of the tr stuff. Most comments are minor.

You should also update `AUTHORS.md`.

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,444 @@
+
+    def estimate_output_error(self, mu):
+        U_mu, pr_err = self.rom.solve(mu, return_error_estimate=True)
+        return pr_err * (2 * self.rom.l2_norm(U_mu) + pr_err)
```
I guess the question is where we implement the output estimators. For linear outputs, this is `self.rom.estimate_output_error(mu)`. We should probably also implement the other estimators in `reductors`.

Line 381 unfortunately does only make sense for the quadratic example, so we should definitely change this. Same for the `DualTRSurrogate`.

TiKeil replied:
I have now changed this to `self.rom.estimate_output_error(mu)` because it makes much more sense to leave it to the `Reductor`

peoe commented on src/pymor/algorithms/bfgs.py:
```diff
@@ -0,0 +1,329 @@
+
+    data = {}
+
+    assert model.output_functional is not None
+    assert model.output_functional.range.dim == 1
```
I was somehow acting under the assumption that the model's output was always handled by the `output_functional` and completely missed the `dim_output` property.

TiKeil commented:
Thanks @peoe, I refactored the code one more time. Apart from the major change that I discuss below, I changed the return value of `output_d_mu` (I need to check in CI whether this has other consequences). I also found a bug regarding the evaluation counts in the TR (ROM counts were not counted correctly).

Regarding the output error estimation in the `TRSurrogate` that @sdrave mentioned: I now decided to leave any sort of error estimation to the `Reductor` and to have as less code as possible in the `TRSurrogate` (only calling `rom.output_error_estimate(mu)` as suggested by @sdrave). In a user-code, this may change and the methods in `TRSurrogate` can be overwritten as we did before. Right now, we only have two `TRSurrogates` that only differ in terms of their `extend`-method (`PrimalOnlyTRSurrogate` where only primal solutions are added to the space and `PrimalAndDualTRSurrogate` with primal and dual in the same space). We could also think about having the `PrimalOnlyTRSurrogate` as the default in the sense that this would be called `TRSurrogate`.

Since we do not have a `Reductor` that can estimate the output error of a non-linear output, the code can currently only be tested for linear outputs. Of course, the TR algorithm already allows for arbitrary outputs with estimation.

As time progresses, I doubt that we will be able to properly include the estimation for bilinear outputs in a `Reductor` until Thursday. Thus, we should think about adding specialized `TRSurrogates` for bilinear outputs or just stick with the linear case as a test case for this release and include the bilinear functionality for the demo (which will be the next PR after the release). @sdrave, what do you think?

Another thing that could soon be added would be a `DWRTRSurrogate` that works with the `DWRCoerciveRBReductor`and internally splits the primal- and dual spaces. However, for this, we would need to compute the gradient of the corrected output (which is at least not trivial).

I think I would try to merge this soon and leave the mentioned WIP to future PRs.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,415 @@
+            with logger.block('Running output checks for TR parameters.'):
+                if current_output + estimate_output < compare_output:
+                    surrogate.extend(mu)
+                    current_fom_output = fom.output(mu)
```
This fom evaluation is not couted, right?

TiKeil replied:
added remarks that this solution is available by using caching.

sdrave reviewed (changes requested):
Generally, I'm fine with the changes. @TiKeil, would you manage to create a PR for the tutorial by Tuesday or so? Since it's only documentation, I think @pmli would agree to still include it into the release. Would you need quadratic outputs for that?

TiKeil commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,415 @@
+                P_h_mu = dual_problem.solve(mu)
+                dual_solutions.append(P_h_mu)
+
+            self.fom_evaluations += 1 + fom.dim_output
```
I now introduced a `CoerciveRB_trust_region` algorithm.

TiKeil commented on src/pymor/models/interface.py:
```diff
@@ -573,7 +573,7 @@ def output_d_mu(self, mu=None, input=None, return_array=False, **kwargs):
             **kwargs
         )
-        return data['output_d_mu']
+        return self.parameters.parse(data['output_d_mu'])
```
removed this from this PR.

TiKeil commented:
> Generally, I'm fine with the changes. @TiKeil, would you manage to create a PR for the tutorial by Tuesday or so? Since it's only documentation, I think @pmli would agree to still include it into the release. Would you need quadratic outputs for that?

@peoe, I added the demo already here and introduced the simplest error estimation for quadratic outputs in a specialized surrogate. This should be good enough until we manage to include it into a `Reductor`.

However, I do not see the tutorial being ready for the release. I added a remark in the tutorial.

To conclude, we should do the following after the release:

- Refactor the tutorial.
- Change the return value of `model.output_d_mu` to a `Parameter`.
- Incorporate quadratic output estimates to `Reductors`.
- Also include DWR (NCD) outputs to the `TRSurrogates`

I will open an issue if this gets merged.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,464 @@
+
+    return mu, data
+
+def CoerciveRB_trust_region(fom, surrogate, parameter_space=None, initial_guess=None, beta=.95, radius=.1,
```
```suggestion
def coercive_rb_trust_region(fom, surrogate, parameter_space=None, initial_guess=None, beta=.95, radius=.1,
```

The other should be in violation of PEP8

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,464 @@
+            self.new_rom = self.new_reductor.reduce()
+
+class QuadraticOutputTRSurrogate(BasicTRSurrogate):
+    """Surrogate for the :func:`trust_region` with only the primal enrichment naive output estimate.
```
```suggestion
    """Surrogate for :func:`trust_region` with only the primal enrichment naive output estimate.
```

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+        return self.rom.output(mu)
+
+    def output_d_mu(self, mu):
+        self.rom_evaluations += 2
```
I guess this is only correct if the adjoint approach is used?

sdrave replied:
Maybe it would make more sense to count the derivative evaluations separately?

TiKeil replied:
I guess the best way would be to have an `EvaluationCounter`-object for all `Models` but this is something for later.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+
+
+class BasicTRSurrogate(TRSurrogate):
+    """Surrogate for the :func:`trust_region` only enriching with the primal solution.
```
```suggestion
    """Surrogate for :func:`trust_region` only enriching with the primal solution.
```

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+            self.new_rom = self.new_reductor.reduce()
+
+class PrimalDualTRSurrogate(TRSurrogate):
+    """Surrogate for the :func:`trust_region` enriching with both the primal and dual solutions.
```
```suggestion
    """Surrogate for :func:`trust_region` enriching with both the primal and dual solutions.
```

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+class QuadraticOutputTRSurrogate(BasicTRSurrogate):
+    """Surrogate for :func:`trust_region` with only the primal enrichment naive output estimate.
+
+    NOTE: This specialized TRSurrogate should soon be made obsolete when the quadratic
+          output estimation is included into a reductor.
```
```suggestion
    .. note:: 
        This specialized TRSurrogate should soon be made obsolete when the quadratic
        output estimation is included into a reductor.
```

sdrave commented on src/pymordemos/tr.py:
```diff
@@ -0,0 +1,114 @@
+                                         + 'where 0 stands for linear and 1 for a quadratic output'),
+    grid_intervals: int = Argument(..., help='Grid interval count.'),
+    training_samples: int = Argument(..., help='Number of samples used for training the reduced basis.')
+):
```
Please add a bit of documentation what this demo does.

sdrave replied:
Otherwise the demo looks nice!

sdrave commented on src/pymordemos/tr.py:
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
```
File needs to be made executable.

TiKeil replied:
I am not sure what you mean?

sdrave commented on src/pymordemos/tr.py:
```diff
@@ -0,0 +1,114 @@
+    report(reference_mu, reference_output, reference_mu, reference_output, reference_data,
+           parameter_space.parameters.parse, descriptor=' of optimization with FOM model')
+    report(bfgs_mu, bfgs_output, reference_mu, reference_output, bfgs_data,
+           parameter_space.parameters.parse, descriptor=' of optimization with fixed ROM model and BFGS method')
```
Would be nice to also see the dimensions of the ROM/number of FOM evaluations here.

TiKeil replied:
I added more information.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+        U, pr_err = self.rom.solve(mu, return_error_estimate=True)
+        cont_est = self.continuity_estimator_output
+        cont = cont_est.evaluate(self.parameters.parse(mu)) if hasattr(cont_est, 'evaluate') else cont_est
+        return cont * (pr_err * (2 * self.rom.l2_norm(U) + pr_err))
```
This assumes that fom/rom actually have an `l2` inner product. I think it would be better to specify the product used to compute the norm in `__init__`. Also it should be documented w.t.r. which norm the continuity constant is estimated.

sdrave commented on src/pymor/algorithms/tr.py:
```diff
@@ -0,0 +1,487 @@
+        if isinstance(output, LincombOperator) and len(output.operators) == 1 and \
+            isinstance(output.operators[0], QuadraticFunctional):
+            # case for quadratic outputs. Should be obsolete soon, see QuadraticOutputTRSurrogate.
+            continuity_estimator_output = output.coefficients[0]
```
Doesn't this completely ignore the norm of the actual functional?

TiKeil replied:
We now use the `MaxThetaParameterFunctional`.

TiKeil commented:
I need another iteration on this later.

sdrave commented:
@TiKeil, I did some more refactoring, as I really did not like the assumptions made in `coercive_rb_trust_region` on the norm of the functional in the quadratic case. Basically, now the user has to provide an appropriate estimate. This then also works for `Models` not coming from our discretization toolkit.

Please take another look. From my side, we can merge this now as is.

sdrave commented:
@pmli, please merge after @TiKeil has had a chance to look at my last changes.

TiKeil reviewed (approved):
Very good changes ! This can be merged IMO.

TiKeil commented:
Thanks @pmli, I just tried to push the same but you were faster ;)
